libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -136,6 +139,9 @@ proc rebalanceMesh(g: GossipSub, topic: string) {.async.} =
           let p = g.peers[id]
           # send a graft message to the peer
           await p.sendGraft(@[topic])
+    
+    # run fanout 
+    g.replenishFanout(topic)
```

Comment:
do this before getGossipPeers or we get g.gossipsub wiped

Reply:
This isn't correct, `fanout` peers are peers that aren't part of the mesh, a `Graft` is sent to announce that the peer is now part of the mesh. Running fanout here doesn't make sense, as it will run it on topics that aren't supposed to be`fanout`.

Reply:
On that note, `fanout` should be replenished on each publish, there is really no way around that, because that is the only time where fanout peers would make sense.

Reply:
Yeah I wasn't sure about this tbh, will check the spec.
The issue it solved was the following:
if we don't do it here basically in getGossipSubPeers (soon following this line)
we will start calling `excl` on `g.gossipsub` and suddenly there are no peers at all anymore in all our tables.. ( and tests would time out)

Reply:
and so the next time we try replenish.. there are no peers to replenish basically

Reply:
will check the spec better about this.. cos I need to understand `getGossipSubPeers` and `excl` too indeed

Reply:
> On that note, `fanout` should be replenished on each publish, there is really no way around that, because that is the only time where fanout peers would make sense.

I see, that explains

Reply:
Still if we don't publish for a while they will get removed from all tables tho.. but don't get me wrong.. maybe that is by spec

Reply:
Yes, for sure check the spec, but as I remember it fan-out peers are dropped after a timeout and if gossip peers are being excluded, they are moved to the mesh. The timeout for fan-out makes sense, because you might have published only one message and then moved on from it, like a broadcast to announce presense. 

So the flow is as follows, peers are added to the gossip table, and moved from there to either fan-out or mesh. Mesh are peers for topics we subscribe to, fan-out are peers for topics that we only publish but don't subscribe to.

Reply:
Yes but the current problem I'm trying to highlight is that the gossip table is cleared every mesh rebalance (so at least every heartbeat), which clear the possibility for any fanout to happen if we didn't publish faster then next heartbeat.
Is this intented?

Reply:
https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub.nim#L205

I really cannot find anything like that in go.. 
https://github.com/libp2p/go-libp2p-pubsub/blob/8945f91465bc3ba418688684a2c29c384cd79647/gossipsub.go#L1614
it just advertises.. also well go has a bit diff tables

Reply:
that `excl` will make the next fanout empty

Reply:
fanout maintenance is in heartbeat in go
https://github.com/libp2p/go-libp2p-pubsub/blob/8945f91465bc3ba418688684a2c29c384cd79647/gossipsub.go#L1436

Reply:
I probably fixed the issue (copied a bit how go does fanout)

Reply:
> https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub.nim#L205
> 
> I really cannot find anything like that in go..
> https://github.com/libp2p/go-libp2p-pubsub/blob/8945f91465bc3ba418688684a2c29c384cd79647/gossipsub.go#L1614
> it just advertises.. also well go has a bit diff tables

huh, that doesn't look right indeed.

Reply:
to clarify, we shouldn't be dropping the peers from the gossip table there, thats a bug.

Reply:
ok thanks for clarifying, I tried removing it but something did not work.. before digging I wanted to make sure with you indeed.

Reply:
fixed that issue